### PR TITLE
Add Rainbow

### LINKED ISSUE #142: Re-Add Round 2 signature scheme Rainbow

### PR DESCRIPTION
**Pull request by mkannwischer**
Resolves #142 

#### Manually checked properties

* [ ] No stringification macros
* [ ] Output-parameter pointers in functions are on the left
* [ ] Negative return values on failure of API functions (within restrictions of FO transform).
* [ ] `const` arguments are labeled as `const`
* [ ] variable declarations at the beginning (except in `for (size_t i=...`)
* Optional:
  * [ ] All integer types are of fixed size, using `stdint.h` types (including `uint8_t` instead of `unsigned char`)
  * [ ] Integers used for indexing are of size `size_t`

### CODE REVIEW

**Comment by mkannwischer** [comment]
There was a bug in the key generation of the cyclic parameter sets of Rainbow. The secret key and signatures in the KATs are not actually valid. 
I've reported this to the authors yesterday and it was already fixed: https://github.com/fast-crypto-lab/rainbow-submission-round2/commit/86fa9ef4339449ae254ca432effeacee1f787d54 . (Thanks Ming-Shing!). 
I've created new testvectors using their code, but we maybe want to keep in mind that they differ from the ones on the NIST website.

**Comment by thomwiggers** [comment]
To fix the issues with Travis, please rebase on top of master (see #196).

**Comment by mkannwischer** [comment]
Finally, this is ready for review. 
There are still many things that could be improved. But I'd appreciate a second opinion on what is the most important. 
Feel free to look at the code with preprocessor conditionals at https://github.com/mkannwischer/rainbow-clean

File: crypto_sign/rainbowIIIc-classic/clean/blas_comm.c [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+/// @param[in]  i         - the index in the vector a.
+/// @return  the value of the element.
+///
+uint8_t PQCLEAN_RAINBOWIIICCLASSIC_CLEAN_gf256v_get_ele(const uint8_t *a, unsigned i) {
+    return a[i];
+}
+
+unsigned PQCLEAN_RAINBOWIIICCLASSIC_CLEAN_gf256v_is_zero(const uint8_t *a, unsigned _num_byte) {
+    uint8_t r = 0;
+    while ( _num_byte-- ) {
```
Comment (thomwiggers) [inline crypto_sign/rainbowIIIc-classic/clean/blas_comm.c]: there are extra spaces:

```suggestion
    while (_num_byte--) {
```

File: crypto_sign/rainbowIIIc-classic/clean/blas_comm.c [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+static void gf256mat_prod_ref(uint8_t *c, const uint8_t *matA, unsigned n_A_vec_byte, unsigned n_A_width, const uint8_t *b) {
+    PQCLEAN_RAINBOWIIICCLASSIC_CLEAN_gf256v_set_zero(c, n_A_vec_byte);
+    for (unsigned i = 0; i < n_A_width; i++) {
+        gf256v_madd(c, matA, b[i], n_A_vec_byte);
+        matA += n_A_vec_byte;
+    }
+}
+
+void PQCLEAN_RAINBOWIIICCLASSIC_CLEAN_gf256mat_mul(uint8_t *c, const uint8_t *a, const uint8_t *b, unsigned len_vec) {
+    unsigned n_vec_byte = len_vec;
```
Comment (thomwiggers) [inline crypto_sign/rainbowIIIc-classic/clean/blas_comm.c]:
```suggestion
    unsigned int n_vec_byte = len_vec;
```

File: crypto_sign/rainbowIIIc-classic/clean/blas_comm.c [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+static void gf256mat_prod_ref(uint8_t *c, const uint8_t *matA, unsigned n_A_vec_byte, unsigned n_A_width, const uint8_t *b) {
+    PQCLEAN_RAINBOWIIICCLASSIC_CLEAN_gf256v_set_zero(c, n_A_vec_byte);
+    for (unsigned i = 0; i < n_A_width; i++) {
+        gf256v_madd(c, matA, b[i], n_A_vec_byte);
+        matA += n_A_vec_byte;
+    }
+}
+
+void PQCLEAN_RAINBOWIIICCLASSIC_CLEAN_gf256mat_mul(uint8_t *c, const uint8_t *a, const uint8_t *b, unsigned len_vec) {
```
Comment (thomwiggers) [inline crypto_sign/rainbowIIIc-classic/clean/blas_comm.c]:
```suggestion
void PQCLEAN_RAINBOWIIICCLASSIC_CLEAN_gf256mat_mul(uint8_t *c, const uint8_t *a, const uint8_t *b, unsigned int len_vec) {
```

File: crypto_sign/rainbowIIIc-classic/clean/blas_comm.c [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+unsigned gf256mat_gauss_elim_ref( uint8_t *mat, unsigned h, unsigned w ) {
+    unsigned r8 = 1;
+
+    for (unsigned i = 0; i < h; i++) {
+        uint8_t *ai = mat + w * i;
+        unsigned skip_len_align4 = i & ((unsigned)~0x3);
+
+        for (unsigned j = i + 1; j < h; j++) {
+            uint8_t *aj = mat + w * j;
+            gf256v_predicated_add( ai + skip_len_align4, !PQCLEAN_RAINBOWIIICCLASSIC_CLEAN_gf256_is_nonzero(ai[i]), aj + skip_len_align4, w - skip_len_align4 );
```
Comment (thomwiggers) [inline crypto_sign/rainbowIIIc-classic/clean/blas_comm.c]:
```suggestion
            gf256v_predicated_add(ai + skip_len_align4, !PQCLEAN_RAINBOWIIICCLASSIC_CLEAN_gf256_is_nonzero(ai[i]), aj + skip_len_align4, w - skip_len_align4);
```

File: crypto_sign/rainbowIIIc-classic/clean/blas_comm.c [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+// choosing the implementations depends on the macros _BLAS_AVX2_ and _BLAS_SSE
+
+#define gf256mat_prod_impl            gf256mat_prod_ref
+#define gf256mat_gauss_elim_impl      gf256mat_gauss_elim_ref
+#define gf256mat_solve_linear_eq_impl gf256mat_solve_linear_eq_ref
+void PQCLEAN_RAINBOWIIICCLASSIC_CLEAN_gf256mat_prod(uint8_t *c, const uint8_t *matA, unsigned n_A_vec_byte, unsigned n_A_width, const uint8_t *b) {
+    gf256mat_prod_impl( c, matA, n_A_vec_byte, n_A_width, b);
+}
+
+unsigned PQCLEAN_RAINBOWIIICCLASSIC_CLEAN_gf256mat_gauss_elim( uint8_t *mat, unsigned h, unsigned w ) {
```
Comment (thomwiggers) [inline crypto_sign/rainbowIIIc-classic/clean/blas_comm.c]:
```suggestion
unsigned PQCLEAN_RAINBOWIIICCLASSIC_CLEAN_gf256mat_gauss_elim(uint8_t *mat, unsigned int h, unsigned int w) {
```

**Pull request merged**